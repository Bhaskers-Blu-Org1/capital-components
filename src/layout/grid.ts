// Based on grid breakpoints http://carbondesignsystem.com/style/grid/design
import cx from 'classnames'

// For now, always assume 12 columns but this could change with media queries.

const breakpointIdentifiers = ['xs', 's', 'm', 'l', 'xl']

export const createGridClass = (args?: {
  isContainer?: boolean
  isFixedColumns?: boolean
  isFluidRows?: boolean
}) => {
  return cx('bx-grid', {
    'bx-container': args && args.isContainer,
    'bx-grid--fixed-columns': args && args.isFixedColumns,
    'bx-grid--fluid-rows': args && args.isFluidRows
  })
}

export type SupportedSizesAsFractions =
  | '1/2'
  | '1/3'
  | '1/4'
  | '1/6'
  | '2/3'
  | '3/4'
  | '5/6'
  | 'all'

export type SupportedSizes =
  | number
  | SupportedSizesAsFractions
  | (number | SupportedSizesAsFractions)[]

const fractionToWhole = {
  '1/6': 2,
  '1/4': 3,
  '1/3': 4,
  '1/2': 6,
  '2/3': 8,
  '3/4': 9,
  '5/6': 10,
  all: 12
}

const determineSize = (size?: number | SupportedSizesAsFractions) => {
  if (size === undefined) {
    return 1
  }
  if (typeof size === 'number') {
    return size
  }
  return fractionToWhole[size]
}

/**
 * Creates classname for a grid column, specified from a given column span (1-12)
 */
export const createColClass = ({
  size,
  height
}: {
  size?: SupportedSizes
  height?: number | number[]
}) => {
  const sizes = Array.isArray(size) ? size.map(determineSize) : [determineSize(size)]
  const heights = Array.isArray(height) ? height : height ? [height] : []

  return cx(
    ...sizes.map((s, index) => `bx-grid__col--${breakpointIdentifiers[index]}--${s}`),
    ...heights.map((s, index) => `bx-grid__height--${breakpointIdentifiers[index]}--${s}`)
  )
}

// These paddings are specific to grid elements.
export type PaddingSingle = 'top' | 'right' | 'bottom' | 'left'

export type Padding = 'none' | 'all' | 'horizontal' | 'vertical' | PaddingSingle | PaddingSingle[]

// Based on css-gridish, build some standard spacing className
export const buildPaddingClasses = (padding?: Padding) => {
  if (!padding || padding === 'none') {
    return undefined
  }
  const paddings = Array.isArray(padding)
    ? padding.map(p => `bx-padding--${p}`)
    : padding === 'all'
      ? ['bx-padding']
      : [`bx-padding--${padding}`]
  return cx(paddings)
}
