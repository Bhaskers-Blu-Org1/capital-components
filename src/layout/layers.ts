// Some utility methods used to determine the layers (or z-indexes) of different components
// Eventually, we can use react context to provide information about
// where to place different layers of the UI
// For now, we're just constants

export enum layers {
  FlyOver = 100,
  Dialog = 200,
  Alert = 300
}
