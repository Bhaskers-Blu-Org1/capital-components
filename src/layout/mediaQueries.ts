import { css } from "emotion";

export interface BreakPointDescriptor<A> {
  base?: A;
  xs?: A;
  s?: A;
  m?: A;
  l?: A;
  xl?: A;
}

// breakpoints as defined in css-gridish.json file.
// We aren't importing it because we don't want to make assumptions
// about what the client can parse. Keep it to js output.
export const breakpoints: Required<BreakPointDescriptor<number>> = {
  base: 0,
  // Numerical values will result in a min-width query
  xs: 576,
  s: 768,
  m: 992,
  l: 1200,
  xl: 1600
};

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type MqDescriptor = Required<BreakPointDescriptor<(arg: string) => string>>;

/** MQ object supports quick media query aware emotion syntax: https://emotion.sh/docs/media-queries
 * It generates the CSS string to be used within other other emotion css strings. Note
 * that the size (m, s, l, etc) indicates the min-width that the css will apply.
 */
export const mqStrings: MqDescriptor = Object.keys(breakpoints).reduce(
  (accumulator, label) => {
    accumulator[label] = cls =>
      `
      @media (min-width:${breakpoints[label]}px) {
        ${cls}${cls.endsWith(";") ? "" : ";"}
      }
    `;
    return accumulator;
  },
  {} as any
);

/** MQ object supports quick media query aware emotion syntax: https://emotion.sh/docs/media-queries
 * It generates the CSS string to be used within other other emotion css strings. Note
 * that the size (m, s, l, etc) indicates the max-width that the css will apply.
 */
export const mqStringsMax: Omit<MqDescriptor, "base"> = Object.keys(breakpoints).reduce(
  (accumulator, label) => {
    // base doesn't make sense for max-widths
    if (label === "base") {
      return accumulator;
    }
    accumulator[label] = cls =>
      `
      @media (max-width:${breakpoints[label]}px) {
        ${cls}${cls.endsWith(";") ? "" : ";"}
      }
    `;
    return accumulator;
  },
  {} as any
);

/** A helper function used to build a string (to be used in emotion) that
 * reflects particular settings for media queries.
 */
export const buildStringForMediaQueries = (
  desc: string | BreakPointDescriptor<string> | undefined,
  predicate: (arg: string) => string
) => {
  if (!desc) return null;

  if (typeof desc === "string") {
    return predicate(desc);
  }
  const mediaQueries = Object.keys(desc).map(key => mqStrings[key](predicate(desc[key])));
  return mediaQueries.join("\n");
};

/** MQ object supports quick media query aware emotion syntax: https://emotion.sh/docs/media-queries
 * It generates actual classnames, as opposed to mqStrings which generates the CSS string. Note
 * that the size (m, s, l, etc) indicates the min-width that the css will apply.
 */
export const mq: MqDescriptor = Object.keys(mqStrings).reduce(
  (accumulator, label) => {
    accumulator[label] = cls => css(mqStrings[label](cls));
    return accumulator;
  },
  {} as any
);

/** MQ object supports quick media query aware emotion syntax: https://emotion.sh/docs/media-queries
 * It generates actual classnames, as opposed to mqStrings which generates the CSS string. Note
 * that the size (m, s, l, etc) indicates the max-width that the css will apply.
 */
export const mqMax: MqDescriptor = Object.keys(mqStrings).reduce(
  (accumulator, label) => {
    accumulator[label] = cls => css(mqStringsMax[label](cls));
    return accumulator;
  },
  {} as any
);
