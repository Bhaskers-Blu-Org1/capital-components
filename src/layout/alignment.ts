import { FlexDirectionProperty } from "csstype";
import { buildStringForMediaQueries, IBreakPointDescriptor } from "../layout/mediaQueries";

export const buildAlignment = (
  direction: FlexDirectionProperty,
  alignment: string | IBreakPointDescriptor<string>
) => {
  return buildStringForMediaQueries(alignment, align => buildAlignmentString(direction, align));
};

export const buildAlignmentString = (direction: FlexDirectionProperty, alignment: string) => {
  const values = alignment.split(" ");
  if (values.length === 1) {
    return `justify-content: ${values[0]}; align-items: ${values[0]}`;
  } else if (values.length === 2) {
    const alignmentProperty = direction === "row" ? "align-items" : "justify-content";
    const inverseAlignmentProperty =
      alignmentProperty === "align-items" ? "justify-content" : "align-items";
    if (values[0] === "horizontal") {
      return `${inverseAlignmentProperty}: ${values[1]}`;
    } else if (values[0] === "vertical") {
      return `${alignmentProperty}: ${values[1]}`;
    }
    return `${alignmentProperty}: ${values[0]}; ${inverseAlignmentProperty}: ${values[1]}`;
  }
  throw new Error("Invalid alignment provided.");
};
