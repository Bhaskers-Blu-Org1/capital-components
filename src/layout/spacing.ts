// based on http://carbondesignsystem.com/style/spacing
import { BreakPointDescriptor, buildStringForMediaQueries } from "./mediaQueries";

// Use for margins within grid elements
export const spacing = {
  "3xs": "0.125rem",
  "2xs": "0.25rem",
  xs: "0.5rem",
  sm: "0.75rem",
  md: "1rem",
  lg: "1.5rem",
  xl: "2rem",
  "2xl": "2.5rem",
  "3xl": "3rem"
};

// Use for vertical layout. For horizontal layout, always use grid elements
export const layout = {
  "2xs": "1rem",
  xs: "1.5rem",
  sm: "2rem",
  md: "3rem",
  lg: "4rem",
  xl: "6rem",
  "2xl": "10rem"
};

export const buildSpacing = (
  padding?: string | BreakPointDescriptor<string>,
  type: "padding" | "margin" = "padding"
) => {
  return buildStringForMediaQueries(padding, pad => buildSpacingFromString(pad, type));
};

export const buildSpacingFromString = (
  padding?: string,
  type: "padding" | "margin" = "padding"
) => {
  if (!padding) return null;

  const values = padding.trim().split(" ");
  if (values.length === 1) {
    return `${type}: ${spacing[values[0]]};`;
  } else if (values.length % 2 === 0 && ["top", "right", "bottom", "left"].includes(values[0])) {
    const resultArr: string[] = [];
    for (let i = 0; i < values.length; i += 2) {
      resultArr.push(`${type}-${values[i]}: ${spacing[values[i + 1]]};`);
    }
    return resultArr.join(" ");
  } else {
    return `${type}: ${values.map(s => spacing[s] || s).join(" ")};`;
  }
};
