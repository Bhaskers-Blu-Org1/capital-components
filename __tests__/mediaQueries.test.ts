import { mq, mqStrings } from "../src/layout/mediaQueries";
import { getStyles } from "jest-emotion";
import * as emotion from "emotion";

it("media queries with basic strings", () => {
  const result = mqStrings.l("background-color: blue");
  expect(result).toEqual(
    expect.stringMatching(/@media \(min-width:1200px\) {\s+background-color: blue;\s+/)
  );
});

it("media queries with classnames", () => {
  mq.l("background-color: blue");
  expect(getStyles(emotion)).toMatchSnapshot();
});
