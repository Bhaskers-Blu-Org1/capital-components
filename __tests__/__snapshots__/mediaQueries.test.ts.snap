// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media queries with classnames 1`] = `"@media (min-width:1200px){.css-oesdlk{background-color:blue;}}"`;
